### PR TITLE
Stats Calculator - 3 April - v1

### PR DESCRIPTION
**Pull request by AntonKrychevskyy**

### CODE REVIEW

File: 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.ts [hunk 1]
```diff
@@ -6,10 +6,17 @@ interface Stats {
...
-    avg: 0,
-  };
+  return nums.reduce(
+    (res, num) => {
+      res.min = res.min > num ? num : res.min;
+
+      return res;
+    },
+    {
+      min: Infinity,
```
Comment (ben-obringer) [inline 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.ts:22]: I'm curious about the use of Infinity. How'd you come up with that solution?

Comment (ben-obringer) [reply 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.ts:22]: Ohh. Nvm now. I didn't realize the object with Infinity was the initial state passed to reduce. That makes sense now. I like that approach especially since you're handling cases where the array is empty.

File: 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.spec.ts [hunk 2]
```diff
@@ -18,5 +18,18 @@ describe('stats calculator', () => {
...
+      expect(calculateStats(FIRST_INPUT).max).toEqual(53);
+    });
+
+    it('should return 100 for [-27, 0, 100, -37, 11, 1] input', () => {
+      expect(calculateStats(SECOND_INPUT).max).toEqual(100);
+    });
+
+    it('should return -2 for [-2, -4, -21, -8, -53, -40] input', () => {
+      expect(calculateStats(THIRD_INPUT).max).toEqual(-2);
+    });
```
Comment (ben-obringer) [inline 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.spec.ts:33]: I see a pattern recurring where you have a case that handles all positive integers, a case that handles integers including 0, and a case that handles all negative integers.

I wonder if it makes sense to make your intentions explicit here. I know Khalil talks about making test cases concrete, but I think you could do that and state the abstract as well. Thoughts?

File: 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.spec.ts [hunk 3]
```diff
@@ -32,4 +32,18 @@ describe('stats calculator', () => {
...
+
+    it('should return 8 for [-27, 0, 100, -37, 11, 1, -1, 37] input', () => {
+      expect(calculateStats(SECOND_INPUT).qty).toEqual(8);
+    });
+
+    it('should return 7 for [-2, -4, -21, -8, -53, -40, -5] input', () => {
+      expect(calculateStats(THIRD_INPUT).qty).toEqual(7);
+    });
+  });
```
Comment (ben-obringer) [inline 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.spec.ts:48]: What about edge cases like a qty. of 0?

**Review by ben-obringer — COMMENTED** [review]
I really like your implementation. I'm going to use reduce instead of a loop : )

File: 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.ts (hunk 1 above)
Comment (AntonKrychevskyy) [reply 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.ts:22]: Well to assign first item value to min and max also smart way)

File: 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.spec.ts (hunk 3 above)
Comment (AntonKrychevskyy) [reply 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.spec.ts:48]: Previously, I added in FizzBuzz the treatment of such edge cases. And gradually up to this task I have become a bit lazy and focused only on the format of test cases, sequence of actions to write them. I must try to bring together all the good practices and strive for perfection.

File: 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.spec.ts (hunk 2 above)
Comment (AntonKrychevskyy) [reply 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.spec.ts:33]: Well as more abstract I am using second wrapping `describe` with definition of tested functionality. Maybe it makes sense to add description of each array to make it more clear on specificity of each case - with positive, mixed or negative values.

Comment (stemmlerjs) [reply 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.spec.ts:33]: Great discussion @ben-obringer & @AntonKrychevskyy. 

Yes, this is great - you've done a good job here of making your test cases explicit examples and wrapping them in the higher level Responsibility (feature) 👍 

The only improvement I'd make here is to parameterize your Examples that exist within your Responsibility Specifications (I'm going to that that to refer to those high-level specifications that you're using `describe` to represent).

Comment (stemmlerjs) [reply 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.spec.ts:33]: Additionally, instead of naming your input data FIRST_INPUT, SECOND_INPUT, THIRD_INPUT, giving them better names to call out the relevant edge cases in each of them is a great way to utilize naming.

For example:

- assignmentRequiredInput
- inputContainingZeros
- inputWithNegativeNumbers
- inputWithoutFiveNumbers

etc.

All potential improvements. Not mad about this solution.

Comment (AntonKrychevskyy) [reply 1_The_Physical/Guess_Point_Level_1_Stateless/2_Arrange_Act_Assert/2_2_Stats_Calculator/src/index.spec.ts:33]: Thank you for suggestions @stemmlerjs. I am not sure that I got idea about parametrizing, but separate input for each edge case with descriptive names definitely nice idea.

Unfortunately I had to pause my walkthrough the course, but hope a few weeks later will get a chance to get back to tasks and course materials.
